Add Code Actions and a few "Quick Fix" actions for warnings [reverted]

Meorge opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/13232 .

This PR adds support for "Code Actions" and "Quick Fixes" in Godot's script editor, for GDScript. This aims to work similarly to [VS Code's Code Actions](https://code.visualstudio.com/docs/editing/refactoring).

Essentially, when a GDScript warning is pushed, a list of Code Actions can optionally be supplied as well. These will be offered to the user via a lightbulb icon that appears in the script editor gutter. The option to suppress the warning with `@warning_ignore` is often provided, and depending on the warning, a more specific fix may also be provided.

Currently the following fixes are provided:

- UNUSED_VARIABLE
  - Add underscore to variable name
  - Remove variable declaration
- UNUSED_PARAMETER
  - Add underscore to parameter name
- MISSING_TOOL
  - Add `@tool`
- GET_NODE_DEFAULT_WITHOUT_ONREADY
  - Add `@onready`
- UNUSED_PRIVATE_CLASS_VARIABLE
  - Remove underscore from class variable name
  - Remove class variable declaration
- UNTYPED_DECLARATION
  - Add type specifier
- INFERRED_DECLARATION
  - Add type specifier
- UNUSED_LOCAL_CONSTANT
  - Add underscore to constant name
  - Remove constant declaration
- ASSERT_ALWAYS_TRUE
  - Remove assert statement
- REDUNDANT_AWAIT
  - Remove "await"
- STATIC_CALLED_ON_INSTANCE
  - Call from type
- MISSING_AWAIT
  - Add "await"

https://github.com/user-attachments/assets/cabba728-afb5-4559-8fa4-aa3b6ad5f102

## TODOs

None at the moment!

## For future PRs

These are things that would be nice to add in future PRs but are **out of scope for this PR:**

- View and run code actions associated with warnings in the warning panel
- Create function stubs to match nonexistent referenced functions
- Create signals to match nonexistent referenced signals
- Create properties to match nonexistent referenced properties(?)

Meorge committed:
GDScript: Add Code Actions and "Quick Fix" actions for warnings/errors

AThousandShips commented on scene/gui/code_edit.h:
```diff
@@ -381,6 +391,16 @@ class CodeEdit : public TextEdit {
+	void set_draw_code_actions(bool p_draw);
+	bool is_draw_code_actions_enabled() const;
+	HashMap<int, Vector<ScriptLanguage::CodeActionOperation>> code_actions;
+	int current_code_action_line; // The last-selected line to use code actions from.
```
```suggestion
	int current_code_action_line = -1; // The last selected line to use code actions from.
```
Or some appropriate default

Meorge commented:
Errors can now also have Code Actions associated with them. For now, this is just utilized so that warnings *treated as errors* can still have their Code Actions appear, but I don't think anything's preventing error-specific Code Actions.

Meorge commented:
Some improvements:

- Quick Fixes are now grouped in the menu by the warning they're addressing.
- Pressing Alt + Enter in the editor brings up the menu.
- A whole bunch more warnings have Quick Fixes:
  - UNUSED_PRIVATE_CLASS_VARIABLE
  - UNTYPED_DECLARATION
  - INFERRED_DECLARATION
  - UNUSED_LOCAL_CONSTANT
  - ASSERT_ALWAYS_TRUE
  - REDUNDANT_AWAIT
  - STATIC_CALLED_ON_INSTANCE
  - MISSING_AWAIT
  - INT_AS_ENUM_WITHOUT_CAST

Unfortunately I've run into what feels like an impasse at the moment. At the point in the analyzer where warnings are pushed, information about the indentation level of the statements seems to have been lost. This makes it difficult to implement code actions that insert text at the same indentation level as the line with the warning, because if a warning is being emitted from a sub-part of the line, we don't know where to insert the new text. Addressing this will likely require storing additional data in the parser/tokenizer nodes. If anyone has better or more efficient ideas on how to approach this, please let me know!

AThousandShips commented on scene/gui/code_edit.cpp:
```diff
@@ -3972,6 +4088,11 @@ CodeEdit::CodeEdit() {
+
+	// Popup for Code Actions
+	code_action_popup = memnew(PopupMenu);
+	code_action_popup->connect("id_pressed", callable_mp(this, &CodeEdit::_on_code_action_id_pressed));
```
```suggestion
	code_action_popup->connect(SceneStringName(id_pressed), callable_mp(this, &CodeEdit::_on_code_action_id_pressed));
```

Meorge commented:
I've made a lot of progress on this! Code Actions now support the LSP. This means that they can be activated from an external editor like VS Code!

In designing this to support the LSP, document edits can now be performed across multiple files simultaneously. This opens the door for other refactoring operations, such as renaming symbols, to be easily implemented using this system.

There is a bit of a difficult point I've run into during this process. Currently, running a multi-document edit operation (such as a Code Action) doesn't apply to the undo/redo stack. It simply opens each file and performs the necessary edits. I think having the ability to undo and redo these multi-document edits will be important; I've found myself attempting to Cmd+Z away a Code Action many times just while testing. Is it standard or acceptable to have a single undo/redo action change the contents of multiple files at once like this?

I've also got to figure out how to properly handle columns and tab characters. Godot seems to treat a tab character as the number of columns specified in the editor settings, so some conversions will be necessary there to get things correctly aligned.

AThousandShips commented on scene/gui/code_edit.h:
```diff
@@ -381,6 +391,17 @@ class CodeEdit : public TextEdit {
+	bool is_draw_code_actions_enabled() const;
+	void show_code_actions(int p_line);
+	HashMap<int, Vector<ScriptLanguage::CodeActionGroup>> code_action_groups;
+	int current_code_action_line = -1; // The last selected line to use code actions from.
```
Should these two really be public?

AThousandShips commented on modules/gdscript/language_server/godot_lsp.h:
```diff
@@ -758,6 +781,11 @@ struct DiagnosticRelatedInformation {
 	String message;
 
+	_FORCE_INLINE_ void load(const Dictionary &p_params) {
+		location.load(p_params["location"]);
```
Are these all guaranteed to be present? Otherwise they should probably be guarded like other work lately

AThousandShips replied:
Are these all guaranteed to be present? Otherwise they should probably be guarded like other work lately

AThousandShips commented on modules/gdscript/gdscript_warning.h:
```diff
@@ -33,6 +33,7 @@
 #include "core/object/object.h"
+#include "core/object/script_language.h"
 #include "core/string/ustring.h"
 #include "core/templates/vector.h"
```
```suggestion
#include "core/object/script_language.h"
```

AThousandShips commented on editor/script/script_editor_plugin.cpp:
```diff
@@ -2759,6 +2761,13 @@ bool ScriptEditor::edit(const Ref<Resource> &p_resource, int p_line, int p_col,
 
+void ScriptEditor::_on_document_edits_applied(const PackedStringArray &p_paths) {
+	// Currently does the same thing as _on_find_in_files_modified_files.
+	print_line("Document edits were applied to files:", p_paths);
```
```suggestion
```

AThousandShips commented on editor/script/script_editor_plugin.cpp:
```diff
@@ -3750,6 +3758,165 @@ void ScriptEditor::_filter_methods_text_changed(const String &p_newtext) {
+				}
+			}
+
+			// Not using an error macro so that the rest of the function continues
```
Any reason `ERR_PRINT` is not appropriate?

Meorge commented:
> It would be nice if this feature did not make the gutter even wider .-.

I definitely agree on the gutter! Looking at VS Code, I realized that they don't actually provide dedicated gutter space for the Quick Fix button, but instead try to find an appropriate position for it around the text. I'm hoping to update it to do something like that instead of the current extra-gutter behavior.

> I tested some quick fixes and none of them worked correctly.

What quick fixes specifically did you try, and how did the results come out? Some of them have been working for me, but others have not. 

> From the comments I assume this is still work in progress? Shouldn't it be marked as draft?

Yep, I think a while back the scope of the PR was smaller and everything included was working better, so I was at least hoping for more specific review. Since then I added several quick fixes, LSP support, and restructured how some things work, so it's back to being on the more WIP side 😅 I'll move this back to a draft to reflect that.

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -3866,6 +3904,71 @@ Error GDScriptLanguage::complete_code(const String &p_code, const String &p_path
+			e.column = pe.start_column;
+			e.message = pe.message;
+
+			for (auto action : fix_ranges_for_code_actions(pe.code_actions, lines).actions) {
```
Same

Meorge commented:
I think I've fixed all of the problems I had written down with the PR, so I'm marking it as ready for review (again)! 🥳

Perhaps most notably on the user side, I removed the extra gutter and replaced it with an icon overlaid on the text itself, when a line has code actions available:

https://github.com/user-attachments/assets/cabba728-afb5-4559-8fa4-aa3b6ad5f102

This button acts similarly to the corresponding control in VS Code. If there is space on the line (i.e., the code there is indented), then the button shows up on that line; if there isn't space on the line, it tries to move the button to either the line above or below.

It's a small visual nitpick, but the button isn't perfectly vertically centered on the line. As best I can tell, it has to do with how Godot defines the `get_caret_draw_pos()` and `get_line_height()` methods. I might spend some more time trying to get that figured out in the coming days, but if anyone happens to know how it'd be approached correctly, I'd love to hear about it!

I'm also feeling a little iffy about the lightbulb icon I'm currently using to indicate code actions. It's a modified version of the PointLight2D/OmniLight3D icon, as I recall. While I like having consistency in the editor for a "lightbulb" icon, I feel like it could stand to be a bit more readable. For reference, here's how the code from my video above looks in VS Code:

<img width="311" height="119" alt="CleanShot 2026-03-19 at 21 23 54@2x" src="https://github.com/user-attachments/assets/5f4ba831-ed96-4e96-9d50-3ddcc14da3b3" />

Anyways, I'm glad to finally have the bulk of the work on this done now. Looking forward to feedback and reviews!

arkology commented:
> this replace-the-line-number behavior

Please note that in master you can click line number to select whole line. With these changes it will be not possible.

arkology commented:
> just need to except the ugliness of the vscode placement

This is how it looks in PyCharm:
<img width="604" height="134" alt="pycharm64_7uevSqgAFV" src="https://github.com/user-attachments/assets/b3998421-41f3-4259-bed2-eb4aae3c01b1" />
<img width="505" height="303" alt="pycharm64_B1e3uXpqET" src="https://github.com/user-attachments/assets/787e35e4-e97e-48d5-9c3b-669dedb48fd3" />

AThousandShips commented on scene/gui/code_edit.cpp:
```diff
@@ -438,6 +444,63 @@ void CodeEdit::_notification(int p_what) {
+		}
+	}
+
+	icon_position.x = round(icon_position.x);
```
```suggestion
	icon_position.x = Math::round(icon_position.x);
```
For all these

Meorge commented:
Related to the previously-discussed issue of the icon being rendered blurry, I found that it looks even worse when in light mode, especially with a weird dark outline around the edge of it:

<img width="309" height="108" alt="CleanShot 2026-03-27 at 17 50 29@2x" src="https://github.com/user-attachments/assets/c8326114-516e-4661-9a35-ad19ad138e29" />
<img width="281" height="96" alt="CleanShot 2026-03-27 at 17 50 50@2x" src="https://github.com/user-attachments/assets/7597cb33-944f-463d-abf9-41124db3b6e8" />

As far as I can tell, I'm rendering this icon the same way that other icons (e.g., for code completion suggestions) are being rendered, and I don't think there's anything particularly different about the icon SVG itself. So at the moment, I'm leaning towards trying to get this PR through as-is, and then opening bug report issues for the strange rendering since that involves a different part of the engine.

kitbdev reviewed (commented):
<img width="507" height="156" alt="Image" src="https://github.com/user-attachments/assets/a1a04594-5130-49c2-a496-ba51775579b4" />

Some issues: 
- The icon is not centered in the line.
- The popup is not at the icon when the icon is clicked. When the shortcut is used, it is good to have at the caret, but it should be at the icon when using the mouse.
- Hovering over the icon should change the mouse cursor to not be the text editing bar.
- Hovering over the icon should show a tooltip saying what it does like "Show Code Actions" and the shortcut.


In other IDEs when there are multiple completions on a line, only the ones valid at the caret location are shown. Not sure if possible to change with current warning info or if it is needed. It's in order so it might be fine.
<img width="468" height="246" alt="Image" src="https://github.com/user-attachments/assets/3e4463b4-319b-414d-827b-5d92d6b0f1e8" />

Meorge commented:
I looked a bit more into the possibility of moving the code action functionality from CodeEdit to CodeTextEditor this evening, and (so far) I don't think it would make sense. Functionality like the icon and tooltip is tied to methods in CodeEdit itself, so accessing those from CodeTextEditor would require work on CodeEdit/TextEdit.

I suppose I could give a shot at another approach to displaying the code actions button, where instead of custom-drawing the icon I just use a Button node that I reposition at the correct location. If that works well, then CodeTextEdit may be able to handle all the logic and leave CodeEdit relatively untouched. But otherwise, at the moment I think the best thing might just be to keep most of the code as-is in CodeEdit, and un-expose things like `document_edits_requested`, *or* work on exposing the whole code actions workflow through CodeEdit so that it can be used by others.

kitbdev commented on doc/classes/EditorSettings.xml:
```diff
@@ -1463,6 +1463,10 @@
 		</member>
+		<member name="text_editor/behavior/code_actions/show_code_actions" type="bool" setter="" getter="">
+			If [code]true[/code], displays a lightbulb button when the cursor is on a line with code actions available.
+			Code actions can perform automated edits to code that address warnings or fix errors.
```
```suggestion
			If [code]true[/code], the script editor displays a lightbulb button when the caret is on a line with code actions available. Clicking the lightbulb button shows available code actions. Code actions can perform automated edits to code that address warnings or fix errors. If [code]false[/code], code actions are only available through the "Show Code Actions" shortcut.
```
Or something.
Added more information about what it does when disabled, caret instead of cursor, and what the lightbulb button does.

Meorge commented:
It's been improved now:

<img width="421" height="314" alt="CleanShot 2026-04-06 at 19 29 53@2x" src="https://github.com/user-attachments/assets/c3f28a8a-4bbf-4855-b4fd-911a0ee71ded" />

I think the underlying cause of this is that the `UNTYPED_DECLARATION` warning for a function's return type encompasses the *whole function* instead of just the function declaration line. Here you can see that it's showing that warning even when I'm hovering over a line in the middle of the function:

<img width="1961" height="627" alt="CleanShot 2026-04-06 at 19 30 38@2x" src="https://github.com/user-attachments/assets/5ee8a282-407d-4438-aeb0-47090cd09ba7" />

So, I think this would need to be addressed with a separate PR that changes the `UNTYPED_DECLARATION` warning to only show for the declaration line.

kitbdev commented on core/object/script_language.h:
This was added recently #117710
Should probably be merged with yours?

Meorge replied:
I've removed my `TextEditOperation` and adopted the existing `TextEdit` struct in its place 🙂

AThousandShips commented on editor/gui/code_editor.cpp:
```diff
@@ -1164,6 +1184,16 @@ void CodeTextEditor::update_editor_settings() {
 
+	// Behavior: Code Actions
+	show_code_actions = EDITOR_GET("text_editor/behavior/code_actions/show_code_actions");
+	if (text_editor->is_connected("draw", callable_mp(this, &CodeTextEditor::_on_text_editor_draw))) {
```
```suggestion
	if (text_editor->is_connected(SceneStringName(draw), callable_mp(this, &CodeTextEditor::_on_text_editor_draw))) {
```
For all here

kitbdev commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -143,10 +148,41 @@ static void get_function_names_recursively(const GDScriptParser::ClassNode *p_cl
+#ifdef TOOLS_ENABLED
+ScriptLanguage::CodeActionGroup fix_ranges_for_code_actions(const ScriptLanguage::CodeActionGroup &p_group, const Vector<String> &p_lines) {
+	// The GDScript analyzer outputs positions within a line as columns, but the
+	// LSP and text editor need them as numbers of characters. Additionally, there
```
This part has changed in https://github.com/godotengine/godot/pull/117710
(It still needs to be fixed for 1 index, but not for tabs)

kitbdev commented on editor/script/script_editor_plugin.cpp:
```diff
@@ -3776,6 +3777,67 @@ void ScriptEditor::_filter_methods_text_changed(const String &p_newtext) {
+		ce->begin_complex_operation();
+		for (const ScriptLanguage::TextEditOperation &text_edit : doc_edit.edits) {
+			ce->select(text_edit.start_line, text_edit.start_col, text_edit.end_line, text_edit.end_col);
+			ce->insert_text_at_caret(text_edit.new_text);
```
These should use remove_text and insert_text so the caret doesn't get changed. 
I think currently if you have multiple selections it would cause issues.

AdriaandeJongh commented:
Tested latest version with a few scripts and it looks great with no remarks. Found one issue where a warning that's on the last line of a file can't be removed:

https://github.com/user-attachments/assets/e64aea10-9aa8-4bf3-adf0-51799d008dbc

Meorge commented:
Thanks for testing! That should now be fixed 🙂 


https://github.com/user-attachments/assets/d063c158-695a-487f-891d-919cb6e0bfd8

AdriaandeJongh reviewed (approved):
LGTM from a usability perspective! 

Didn't look at the code.

Meorge commented:
While [writing a response to a discussion thread](https://github.com/godotengine/godot-proposals/discussions/14834#discussioncomment-16894668) I discovered that the `INFERRED_DECLARATION` code action is broken (doesn't replace `:=` with `=`) so I'll need to get that fixed soon. It's now on the TODO list.

Meorge commented:
`INFERRED_DECLARATION` issue is fixed, and inserting type specifiers now works for `UNTYPED_DECLARATION` too!

https://github.com/user-attachments/assets/c7c0d84a-57be-44f9-8927-3c16c4b05851

kitbdev commented on editor/gui/code_editor.cpp:
```diff
@@ -960,6 +965,16 @@ void CodeTextEditor::_text_editor_gui_input(const Ref<InputEvent> &p_event) {
+
+	if (show_code_actions && mm.is_valid()) {
+		if (_get_code_action_button_inline_rect().has_point(mm->get_position()) && get_code_actions_for_line(text_editor->get_caret_line()).size() > 0) {
+			text_editor->set_default_cursor_shape(Control::CURSOR_POINTING_HAND);
```
This should stop the symbol tooltip on hover timer too somehow. If this is called first `accept_event` should work, but if not then CodeEdit should have some public way to cancel it.

<img width="358" height="102" alt="Image" src="https://github.com/user-attachments/assets/572755e6-e753-4ced-b23a-dc6df563c6e5" />

Meorge replied:
I'm implementing checks in `CodeTextEditor::_text_editor_gui_input()` that disable the symbol tooltip when the user is pointing at the code actions button, and re-enable it otherwise. The symbol tooltip now disappears before the code actions tooltip appears.

https://github.com/user-attachments/assets/66aaf21b-0908-40f7-bb75-1cf5f40fca1a

kitbdev commented on editor/gui/code_editor.cpp:
```diff
@@ -1945,15 +1979,135 @@ void CodeTextEditor::update_toggle_files_button() {
+	const int row_height = text_editor->get_line_height();
+	Size2 icon_size(row_height, row_height);
+
+	Point2 icon_position = text_editor->get_rect_at_line_column(line, 0).position;
```
When the line is out of view (scroll up/down too much) then the code actions button should not show at all. Currently it shows at the top of the TextEdit.
So this should check when `get_rect_at_line_column` returns -1, and return early with a Rect that shows its invalid, and then in `_on_text_editor_draw`, it can detect it and not draw it. 
The other uses of `_get_code_action_button_inline_rect` could check too, or the returned size could just be 0,0 since they just check `has_point`.

kitbdev commented on editor/gui/code_editor.cpp:
```diff
@@ -1945,15 +1979,135 @@ void CodeTextEditor::update_toggle_files_button() {
+		}
+
+		// Check if line below has space.
+		else if (line <= text_editor->get_line_count() - 1 && (text_editor->get_indent_level(line + 1) >= 4 || text_editor->get_line_width(line + 1) == 0)) {
```
I'm getting error spam when on the last line and the above line is taken, this check seems off.
Delete the last line and don't save to reproduce (since a new line is automatically added).

<img width="761" height="202" alt="Image" src="https://github.com/user-attachments/assets/d82da7c8-927f-41f3-ac45-87dc4d7ed6a9" />

Also there is not guaranteed to be space underneath it (`text_editor/behavior/navigation/scroll_past_end_of_file` may be false). Not sure what else to do in that case though, so its probably fine. 

This can probably be combined with the next else (haven't tested):

```suggestion
		else if (line == text_editor->get_line_count() - 1 || (text_editor->get_indent_level(line + 1) >= 4 || text_editor->get_line_width(line + 1) == 0)) {
```

Meorge replied:
Tested on my end and the fix seems to work well, thanks!

kitbdev commented on editor/gui/code_editor.cpp:
```diff
@@ -1945,15 +1979,135 @@ void CodeTextEditor::update_toggle_files_button() {
+
+		// Check if line below has space.
+		else if (line <= text_editor->get_line_count() - 1 && (text_editor->get_indent_level(line + 1) >= 4 || text_editor->get_line_width(line + 1) == 0)) {
+			icon_position.y += row_height;
```
```suggestion
			icon_position.y += row_height * get_line_wrap_amount(line);
```
This doesn't take into account word wrapping.
You should be able to multiply `row_height` by `get_line_wrap_amount(line)` to move it down.
(Same in next condition.)


<img width="577" height="178" alt="Image" src="https://github.com/user-attachments/assets/76d398ea-3a56-471d-a76a-2e929a86b577" />

kitbdev commented on editor/gui/code_editor.cpp:
```diff
@@ -1945,15 +1979,135 @@ void CodeTextEditor::update_toggle_files_button() {
+	const int row_height = text_editor->get_line_height();
+	Size2 icon_size(row_height, row_height);
+
+	Point2 icon_position = text_editor->get_rect_at_line_column(line, 0).position;
```
This also doesn't consider horizontal scrolling. I'm not sure if we want to support it, but it may be useful if there is a lot of nesting. Or we could wait and see if there is demand for it.

If we do want to support it, then:
The column in `get_rect_at_line_column`, would have to be a position always on screen (it gets overridden to after the gutters anyway), maybe based on the horizontal scroll value.

It would also affect the logic to find an empty space for it, so instead of indent checks we could use `get_line_column_at_pos(icon_position, false, false)` to determine what the icon would cover, and if it is whitespace or returns (-1,-1) then the position is good.

Meorge replied:
I think for now, it may make sense to hold off. As of right now, the button disappears when it goes offscreen due to horizontal scrolling; this matches the behavior of it disappearing when it goes offscreen due to vertical scrolling.

kitbdev commented on doc/classes/EditorSettings.xml:
```diff
@@ -1493,6 +1493,9 @@
 			The space to add between lines (in pixels). Greater line spacing can help improve readability at the cost of displaying fewer lines on screen. Negative values allow for even more compact text, but may look broken with certain fonts.
 		</member>
+		<member name="text_editor/behavior/code_actions/show_code_actions_button" type="bool" setter="" getter="">
+			If [code]true[/code], the script editor displays a lightbulb button when the caret is on a line with code actions available. Clicking the lightbulb button shows available code actions. Code actions can perform automated edits to code that address warnings or fix errors. If [code]false[/code], code actions are only available through the "Show Code Actions" shortcut.
```
`Code actions can perform automated edits to code that address warnings or fix errors`
Code actions shouldn't only be associated with warnings or errors. They should be able to exist independently, so they can be provided even when there is no warning. Like in VSCode they can be used for refactoring, for example to extract selected text to new function, or invert an if condition. 
I'm not familiar with the GDScript side, but it should be possible to add them in other ways as well.
Not that it has to be done in this PR, so this could be changed later.

Meorge replied:
I definitely agree, and would love to have more general refactoring code actions in the future!

I think this is a case where we need to ensure we have the terminology correct. If my memory serves, "code actions" is a more general term that refers to an automated change to the code and a "quick fix" is a type of code action that specifically targets a diagnostic? I can double-check the VS Code and LSP documentation later to confirm, so that we can be consistent with them.

kitbdev reviewed (commented):
> Multi-document undo redo (https://github.com/godotengine/godot/pull/114168#issuecomment-3830464005)

Do any of the current Code Actions work on multiple documents at once? If none of the current actions need it could be done later.
I tested how VSCode handles multiple document undo redo, and they work as long as the files are at the same undo level as when the action was done (a warning shows if any documents were changed and couldn't do the undo/redo with the rest). We don't have anything for it yet, but it makes sense to me to have it work across multiple documents for code actions.
I have a plan of how the undo/redo can work with multiple documents:
1. When a multi-document edit happens, save the file names and their undo versions (`TextEdit::get_version()`) in a new undo stack on ScriptEditor
2. Intercept undo and redo action handling before the TextEdit gets it (either with existing `_text_edit_gui_input` or a new one in ScriptEditor). On undo/redo, if the TextEdit current undo version matches the undo version in the current item in the stack, then we are going to do a multi-document undo/redo. If it does not match, don't accept the event and let TextEdit get the undo/redo like normal.
3. For the multi-document undo/redo, accept the undo/redo event, do an undo/redo in the current TextEdit, and check all the other files saved in the current undo stack item. If they are open and if their current undo version matches then undo/redo in those TextEdits too. And then move the stack position to the previous/next item.
	1. If the undo version does not match, warn the user about all the files that failed.
4. When a document is closed, go through all items in the undo stack and mark that file if it has it so it can no longer be a part of the multi-document undo/redo. This is because the TextEdit undo/redo stack is not saved when its closed.

Meorge commented:
Thanks so much for this thorough review! 😄 

> Do any of the current Code Actions work on multiple documents at once? If none of the current actions need it could be done later.

To the best of my memory, and from skimming the changes, no current code actions work on multiple documents at once. However, the format for storing code action information defined by the LSP (which I've tried to preserve as much as possible for the internal Godot implementation) _does_ allow for edits to multiple documents to be specified as part of the overall code action operation.

The current code for performing the document edits (including multiple documents at once, if it is passed such an edit request) can be found at [`ScriptEditor::_on_document_edits_requested()`](https://github.com/Meorge/godot/blob/f6edc05ed534ba7c9505a5a30acca05edff1c591/editor/script/script_editor_plugin.cpp#L3780). It opens a script editor for every document that's part of the overall edit operation, then performs all of the edits for each document in the corresponding editor.

For now, since no multi-document code actions are defined, I suppose we could hardcode Godot to only perform the edits for the first (and only) document in the operation. Then once we have multi-document edit operations, we can do a follow-up PR that implements a more robust system like the one you described. If we did this, we could also add a warning/error message to the `_on_document_edits_requested()` method that triggers if a multi-document edit is attempted. While I think the current implementation would work "well enough" for multi-document edits, it hasn't been tested for that. So I wouldn't want someone to try implementing one and assuming it would definitely work 😅

---

I may not be able to address all of these comments immediately, but I'll try to get to them ASAP. Thank you again for this feedback!

Meorge commented:
Suggestions should be all implemented now, I think!

As noted in the comment thread, for now I haven't added support for horizontal scrolling. Just like the button isn't displayed when it scrolls off-screen vertically, it also vanishes when it moves off the screen horizontally. The logic for correctly placing it sounds like it could get to be gnarly, and I wouldn't expect it to come up as an issue very often. However, if this PR is merged and then we get complaints about the button not being accessible for long lines with word wrap turned off, we can certainly revisit it.

kitbdev commented on editor/gui/code_editor.cpp:
```diff
@@ -1945,15 +1981,137 @@ void CodeTextEditor::update_toggle_files_button() {
+	const int line = text_editor->get_caret_line();
+	const int indent_level = text_editor->get_indent_level(line);
+
+	const int row_height = text_editor->get_line_height();
```
When the line spacing is increased, the icon is big enough to overlap text.

eg. with `text_editor/appearance/whitespace/line_spacing` set to 16:

<img width="457" height="135" alt="Image" src="https://github.com/user-attachments/assets/19c565ef-d306-466e-8725-34c263207a1a" />

Maybe row_height should exclude the line spacing?

Meorge replied:
Done, here's how it looks now on my end with different line spacing!

<img width="325" height="118" alt="CleanShot 2026-06-09 at 16 14 26@2x" src="https://github.com/user-attachments/assets/dc9bbd4a-8f77-47d8-87a6-7c0e3eb8378c" />

<img width="450" height="263" alt="CleanShot 2026-06-09 at 16 15 07@2x" src="https://github.com/user-attachments/assets/6ab4c8a9-7cfc-45b3-af3b-31415a281a6d" />

kitbdev commented on editor/gui/code_editor.cpp:
```diff
@@ -937,6 +937,11 @@ void CodeTextEditor::_text_editor_gui_input(const Ref<InputEvent> &p_event) {
 	Ref<InputEventMouseButton> mb = p_event;
 
 	if (mb.is_valid()) {
+		if (show_code_actions && mb->is_pressed() && _get_code_action_button_inline_rect().has_point(mb->get_position())) {
```
This should check for left mouse button, since using scroll wheel a couple times causes the popup to show and then go blank for some reason, and the right mouse button doesn't open it correctly (conflict with context menu or something?)

kitbdev commented on editor/gui/code_editor.cpp:
```diff
@@ -960,6 +965,18 @@ void CodeTextEditor::_text_editor_gui_input(const Ref<InputEvent> &p_event) {
 
+	Ref<InputEventMouseMotion> mm = p_event;
+
+	if (show_code_actions && mm.is_valid()) {
```
This should check `mm->get_button_mask().is_empty()` to make sure the cursor doesn't change when selecting and the mouse moves over it.

kitbdev commented on editor/script/script_editor_plugin.cpp:
```diff
@@ -3776,6 +3777,75 @@ void ScriptEditor::_filter_methods_text_changed(const String &p_newtext) {
+	// Get all of the files that need to be opened.
+	// If a file isn't currently opened, then open it and add it to a list.
+	HashMap<String, TextEditorBase *> files_and_editors;
+	Vector<String> files_to_open_editor_for;
```
Unused

kitbdev commented on editor/gui/code_editor.cpp:
```diff
@@ -1945,15 +1981,137 @@ void CodeTextEditor::update_toggle_files_button() {
+	}
+
+	Rect2 line_rect = text_editor->get_rect_at_line_column(text_editor->get_caret_line(), 0);
+	Point2 pos = text_editor->get_global_transform_with_canvas().xform(line_rect.position);
```
If the shortcut is used, then this should be at the caret position including the column especially since it could be scrolled horizontally. It causes the popup to show at the top left corner.

Also if wrapping is enabled, the popup covers the line, so when clicking it should probably use the actual icon position, or at least take into account the line wraps.

Meorge replied:
I have most of this implemented now: using the shortcut places the menu at the caret position (including column), and when clicking the button it instead finds its position based on the button's position.

I've run into a difficult situation with preventing the top-left-corner-popup problem. In VS Code, I see that the editor simply moves the viewport such that the caret is visible again, and then places the popup at the now-visible caret. I tried to do the same thing here:

```cpp
Rect2 line_rect;
int l = text_editor->get_caret_line();
int c = text_editor->get_caret_column();

if (!text_editor->is_caret_visible()) {
	text_editor->adjust_viewport_to_caret();
}
line_rect = text_editor->get_rect_at_line_column(l, c);
```
The viewport shifts to make the caret visible, as expected, but the popup still appears at the top-left corner. I'm fairly sure this is because the data that TextEdit uses for `get_rect_at_line_column()` isn't updated until the frame/redraw following `adjust_viewport_to_caret()`. 

Does this sound like a likely cause to you? If so, do you know if there's an easy way to force it to recalculate those values? If not, I may be able to move that code into a dedicated function and call it from `adjust_viewport_to_caret()` so that it's immediately available, but first I'd like to ensure there's not a more straightforward solution I'm missing 😅

Edit: Looking closer at the `NOTIFICATION_DRAW` code for TextEdit, I'm not quite as sure now that it'd be possible/feasible to extract the code for recalculating the rect information...

kitbdev replied:
Yeah, `get_rect_at_line_column` has a lot of problems it really should not be dependent on the draw call.
The easiest workaround for now is probably to defer after adjusting.
Eventually `get_rect_at_line_column` should be replaced with a new method that doesn't have its issues.

Meorge replied:
Thanks for the response! Looks like the approach of separating the popup code into a separate function that I defer is working 😄

kitbdev commented on editor/gui/code_editor.h:
```diff
@@ -229,6 +229,24 @@ class CodeTextEditor : public VBoxContainer {
+	virtual String get_tooltip(const Point2 &p_pos) const override;
+
+	bool show_code_actions = true;
+	bool code_action_button_hovered = false;
```
unused

kitbdev commented on editor/gui/code_editor.h:
```diff
@@ -229,6 +229,24 @@ class CodeTextEditor : public VBoxContainer {
+	Vector<ScriptLanguage::CodeActionGroupWithDiagnostics> get_code_actions_for_line(int p_line) const;
+
+	Vector<ScriptLanguage::CodeActionGroupWithDiagnostics> code_actions;
+	Vector<ScriptLanguage::CodeActionOperation> ungrouped_current_code_actions;
```
This one isn't passed around, so it should probably use LocalVector
```suggestion
	LocalVector<ScriptLanguage::CodeActionOperation> ungrouped_current_code_actions;
```

kitbdev commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -277,11 +279,110 @@ void GDScriptParser::push_warning(const Node *p_source, GDScriptWarning::Code p_
+	action_group.actions.append_array(p_code_actions);
+
+#ifdef TOOLS_ENABLED
+	// Create the ignore code action.
```
The `Ignore "MISSING_TOOL"` does not work.
I get errors:
```
  ERROR: scene\gui\text_edit.cpp:4372 - Index p_from_column = -1 is out of bounds (text[p_from_line].length() + 1 = 20).
  ERROR: scene\gui\text_edit.cpp:4358 - Index p_column = -1 is out of bounds (text[p_line].length() + 1 = 20).
```
Same for Ignore `redundant_static_unload`.

Meorge replied:
Fixed! 😄 

https://github.com/user-attachments/assets/ea010d46-fdd9-421e-b537-644476fdfa68

kitbdev commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -6152,7 +6507,7 @@ void GDScriptAnalyzer::is_shadowing(GDScriptParser::IdentifierNode *p_identifier
 		if (base_class != nullptr) {
 			if (base_class->has_member(name)) {
-				parser->push_warning(p_identifier, GDScriptWarning::SHADOWED_VARIABLE, p_context, p_identifier->name, base_class->get_member(name).get_type_name(), itos(base_class->get_member(name).get_line()));
+				parser->push_warning(p_identifier, GDScriptWarning::SHADOWED_VARIABLE, {}, p_context, p_identifier->name, base_class->get_member(name).get_type_name(), itos(base_class->get_member(name).get_line()));
```
Using `Ignore SHADOWED_VARIABLE` when it is a parameter in a function causes it to be inserted after the f in func:

<img width="474" height="65" alt="Image" src="https://github.com/user-attachments/assets/a4d1eb44-9d1d-4fa0-88db-caf362f8eafa" />

It works fine when not a parameter.

Meorge replied:
Fixed!


https://github.com/user-attachments/assets/3b533b45-06c8-4361-b14f-d1a1fc7da43a

kitbdev commented on editor/gui/code_editor.cpp:
After any text is changed, we should force clear the code actions.
This is because it currently depends on idle parse to clear, but it may be invalid already, and trying to apply an invalid code action can cause errors.
And the idle parse delay is configurable in the settings, so there may be plenty of time to open the code actions popup and apply one to an invalid state.

kitbdev commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -277,11 +279,110 @@ void GDScriptParser::push_warning(const Node *p_source, GDScriptWarning::Code p_
+	action_group.actions.append_array(p_code_actions);
+
+#ifdef TOOLS_ENABLED
+	// Create the ignore code action.
```
Adding a warning ignore annotation puts it after other annotations, even if they were all on the same line. I would expect it to put it on a new line before them instead. 

Example: Ignoring `onready_with_export`:
From 
```gdscript
@onready @export var onready_with_export = 1
```
to:
<img width="481" height="91" alt="Image" src="https://github.com/user-attachments/assets/84fb226a-6f84-4d3c-8dec-be88105d6e4a" />

It does work as is, but it would probably be nicer if it was before the other annotations IMO.

kitbdev commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -277,11 +279,110 @@ void GDScriptParser::push_warning(const Node *p_source, GDScriptWarning::Code p_
+		}
+	}
+
+	ignore_op.new_text = vformat("@warning_ignore(\"%s\")\n%s", warning_name.to_lower(), leading_whitespace);
```
A difference from the existing warning ignore system (`ScriptTextEditor::_warning_clicked`) is that this always adds a new `@warning_ignore()` line, but before it tried to detect if there was already a warning ignore and add the warning to the list.
```gdscript
func test_unsafe_void_return() -> void:
	# using warning panel:
	#@warning_ignore("unsafe_method_access", "unsafe_void_return")
	# using code actions:
	@warning_ignore("unsafe_method_access")
	@warning_ignore("unsafe_void_return")
	return variant_func().f()
```

It would be nice if this could do the same and add to a previous `warning_ignore`.

(In the future the current warning panel ignore system should be replaced with this)

kitbdev commented on editor/gui/code_editor.cpp:
```diff
@@ -1945,15 +1981,137 @@ void CodeTextEditor::update_toggle_files_button() {
+	if (action_index == 0) {
+		return;
+	}
+
```
The popup menu should automatically focus the first item, so you don't need to hit down on the keyboard first.
I think like this should work:
```suggestion

code_action_popup->set_focused_item(0);
```

kitbdev commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -277,11 +279,110 @@ void GDScriptParser::push_warning(const Node *p_source, GDScriptWarning::Code p_
+	action_group.actions.append_array(p_code_actions);
+
+#ifdef TOOLS_ENABLED
+	// Create the ignore code action.
```
Ignoring `unreachable_pattern` messes up the indentation

```gdscript
func test_unreachable_pattern():
	match 1:
		_:
			print(0)
		1:
			print(1)
```
to

<img width="427" height="184" alt="Image" src="https://github.com/user-attachments/assets/9a176d6c-94d5-4279-87da-c5933f07f65f" />

Meorge replied:
Fixed! 😄 


https://github.com/user-attachments/assets/a9eb846f-93ea-46ae-87fa-5c5c33642998

kitbdev reviewed (commented):
> So, I think this would need to be addressed with a separate PR that changes the UNTYPED_DECLARATION warning to only show for the declaration line. https://github.com/godotengine/godot/pull/114168/#issuecomment-4196115803

Similar to `UNTYPED_DECLARATION`, warnings that apply to the entire class highlight the entire class over LSP, such as `redundant_static_unload` and `missing_tool`. These should also probably be limited to just the class declaration line since highlighting the entire file isn't helpful. Also for a future PR.
The script editor has a custom limit to make sure the warnings aren't too long which is why it only shows on the LSP side (`ScriptTextEditor::_update_background_color`).

---
Tested for lots of warnings in editor and over LSP, it mostly works well.

Class-level warnings caused code actions to not work at all over LSP, (like `missing_tool` and `redundant_static_unload`) it just says no code actions available. If those warnings are not present, then it works fine for other warnings.

Finishxx commented:
I think I found a bug with the MISSING_TOOL quick-fix. I pressed Enter on the `Ignore "MISSING TOOL"` action, but nothing happened. Please see [Recording](https://github.com/user-attachments/assets/bec17948-5971-44f7-9961-ba9839ea0bf1)

Meorge commented:
~~@Finishxx Can you include any errors that might be appearing in the Output panel and/or command line window? It sounds like you may be running into the same issue as kitbdev here: https://github.com/godotengine/godot/pull/114168#discussion_r3383032295~~

This issue has been fixed, and will be pushed shortly!

Meorge commented:
> Similar to UNTYPED_DECLARATION, warnings that apply to the entire class highlight the entire class over LSP, such as redundant_static_unload and missing_tool. These should also probably be limited to just the class declaration line since highlighting the entire file isn't helpful. Also for a future PR.

@kitbdev I've submitted issues for both of these (https://github.com/godotengine/godot/issues/120174 and https://github.com/godotengine/godot/issues/120175) and should be able to work on them in the coming days, unless someone else gets to them first! 😄

kitbdev commented on editor/gui/code_editor.cpp:
```diff
@@ -1977,15 +2016,166 @@ void CodeTextEditor::update_toggle_files_button() {
 	ED_SHORTCUT("script_editor/zoom_out", TTRC("Zoom Out"), KeyModifierMask::CMD_OR_CTRL | Key::MINUS);
 	ED_SHORTCUT_ARRAY("script_editor/reset_zoom", TTRC("Reset Zoom"),
 			{ int32_t(KeyModifierMask::CMD_OR_CTRL | Key::KEY_0), int32_t(KeyModifierMask::CMD_OR_CTRL | Key::KP_0) });
+	ED_SHORTCUT("script_editor/code_actions", TTRC("Show Code Actions"), KeyModifierMask::ALT | Key::ENTER);
```
Recently, `alt+enter` has been taken as a fullscreen toggle shortcut.
I think it always has priority, so we need to find a different shortcut.
IDK what though every other editor uses `alt+enter` or `ctrl+.` (which is already bound to next breakpoint)
`alt+.` maybe?

Meorge replied:
I put together a table of several IDEs, the shortcuts they have for the equivalent of our "Show Code Actions" operation, and what existing shortcuts Godot has bound to that at the moment (via searching in Editor Settings's Shortcuts tab):

| IDE           | Shortcut               | Taken by Godot                                        |
| ------------- | ---------------------- | ----------------------------------------------------- |
| Eclipse       | Cmd + 1                | Spatial Editor > 1 Viewport                           |
| Notepad++     | None                   | n/a                                                   |
| Qt Creator    | Alt + Enter            | Common > UI Toggle Fullscreen                         |
| Rider         | Ctrl + Alt + Shift + T | No                                                    |
| Sublime Text  | None                   | n/a                                                   |
| VS Code       | Cmd + .                | Stop Running Project                                  |
| Visual Studio | Cmd + . or Alt + Enter | Stop Running Project or Common > UI Toggle Fullscreen |
| Xcode         | Cmd + '                | Canvas Item Editor > Toggle Grid                      |
| Zed           | Cmd + .                | Stop Running Project                                  |

(Note that these are Mac shortcuts; for Windows, replace Cmd with Ctrl)

It looks from this like the following could work:
- `Cmd + 1`
- `Ctrl + Alt + Shift + T` (or maybe a different letter)
- `Cmd + '`

Additionally, `Alt + .` as you suggested does not appear to be bound to anything. However, at least on macOS, `Alt + .` currently types the greater-than-or-equal-to (`≥`) sign, whereas `Cmd + '` does not type anything. So for now, I think I will use `Cmd + '`.

kitbdev commented on editor/gui/code_editor.cpp:
```diff
@@ -1977,15 +2016,166 @@ void CodeTextEditor::update_toggle_files_button() {
+	const int line = text_editor->get_caret_line();
+	const int indent_level = text_editor->get_indent_level(line);
+
+	const int line_spacing = EDITOR_GET("text_editor/appearance/whitespace/line_spacing");
```
This should use the theme constant from the text edit, not the editor setting itself.
Like `text_editor->get_theme_constant("line_spacing")`.

kitbdev commented on editor/gui/code_editor.cpp:
```diff
@@ -963,6 +968,18 @@ void CodeTextEditor::_text_editor_gui_input(const Ref<InputEvent> &p_event) {
+	Ref<InputEventMouseMotion> mm = p_event;
+
+	if (show_code_actions && mm.is_valid() && mm->get_button_mask().is_empty()) {
+		if (_get_code_action_button_inline_rect().has_point(mm->get_position()) && get_code_actions_for_line(text_editor->get_caret_line()).size() > 0) {
```
```suggestion
		if (_get_code_action_button_inline_rect().has_point(mm->get_position()) && !get_code_actions_for_line(text_editor->get_caret_line()).is_empty()) {
```

kitbdev commented on editor/gui/code_editor.cpp:
```diff
@@ -1977,15 +2016,166 @@ void CodeTextEditor::update_toggle_files_button() {
+}
+
+void CodeTextEditor::_on_text_editor_draw() const {
+	if (!code_action_dropdown_icon.is_valid() || get_code_actions_for_line(text_editor->get_caret_line()).size() <= 0) {
```
```suggestion
	if (!code_action_dropdown_icon.is_valid() || get_code_actions_for_line(text_editor->get_caret_line()).is_empty()) {
```

kitbdev commented on editor/gui/code_editor.cpp:
```diff
@@ -1977,15 +2016,166 @@ void CodeTextEditor::update_toggle_files_button() {
+			}
+		}
+
+		if (filtered_group.actions.size() > 0) {
```
```suggestion
		if (!filtered_group.actions.is_empty()) {
```

kitbdev commented on editor/gui/code_editor.cpp:
```diff
@@ -1977,15 +2016,166 @@ void CodeTextEditor::update_toggle_files_button() {
+
+		// If this is the last line, or the line below has space for the button, we can put it there.
+		else if (line == text_editor->get_line_count() - 1 || (text_editor->get_indent_level(line + 1) >= 4 || text_editor->get_line_width(line + 1) == 0)) {
+			icon_position.y += row_height * (1 + text_editor->get_line_wrap_count(line));
```
These 2 `row_heights` should not be affected by line spacing and should instead use the full `text_editor->get_line_height()`

kitbdev commented on core/object/script_language.h:
```diff
@@ -419,6 +502,8 @@ class ScriptLanguage : public Object {
 	virtual void add_named_global_constant(const StringName &p_name, const Variant &p_value) {}
 	virtual void remove_named_global_constant(const StringName &p_name) {}
 
+	virtual Error get_code_actions(const String &p_code, const String &p_path, Vector<CodeActionGroupWithDiagnostics> *r_actions) { return ERR_UNAVAILABLE; }
```
We probably want to put this in EditorLanguage instead.

Meorge replied:
Thank you for the suggestion! Looking at that class and its documentation description, it does sound like it'd be a good fit for it.

I tried doing migrating the relevant code over to that class, but ran into some issues with includes and forward declarations. I _think_ that the ScriptLanguage Warning and ScriptError classes still need to have references to the CodeActionGroup struct, since the code actions are passed as belonging to those warnings/errors. However, because they are members of EditorLanguage, and EditorLanguage is forward-declared in ScriptLanguage, the compiler won't let me reference those nested types. If I try to replace the EditorLanguage forward-declaration with a full include, then I start getting other forward-declaration issues instead...

Right now, I'm thinking it may be possible to put the code action-specific structs into their own dedicated header file, and reference them as necessary from both ScriptLanguage and EditorLanguage. But I could see this running into the same issues with forward declarations and recursive references. Do you have any thoughts on how it would be best to structure this to avoid these problems?

Meorge replied:
I took a fresh look at this and decided to just move the `get_code_actions()` method itself over, keeping other components of the code actions system (such as the various structs) in place where they were.

kitbdev commented on modules/gdscript/gdscript_parser.h:
```diff
@@ -1488,6 +1591,12 @@ class GDScriptParser {
 		list = node;
 
+#ifdef TOOLS_ENABLED
+		if (nodes_in_progress.size() > 0) {
```
```suggestion
		if (!nodes_in_progress.is_empty()) {
```

kitbdev commented on editor/gui/code_editor.cpp:
```diff
@@ -1977,15 +2016,166 @@ void CodeTextEditor::update_toggle_files_button() {
+}
+
+String CodeTextEditor::get_tooltip(const Point2 &p_pos) const {
+	if (show_code_actions && _get_code_action_button_inline_rect().has_point(p_pos) && get_code_actions_for_line(text_editor->get_caret_line()).size() > 0) {
```
```suggestion
	if (show_code_actions && _get_code_action_button_inline_rect().has_point(p_pos) && !get_code_actions_for_line(text_editor->get_caret_line()).is_empty()) {
```

AThousandShips commented:
I'd suggest marking as draft until it is ready to be merged/reviewed again

Meorge commented:
Sounds like a good idea, currently with the process of moving some of the code from ScriptLanguage to EditorLanguage it's a bit messy 😅

kitbdev commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -2266,7 +2466,39 @@ void GDScriptAnalyzer::resolve_variable(GDScriptParser::VariableNode *p_variable
+			append_op_de.file_path = parser->script_path;
+			append_action.document_edits.append(append_op_de);
+
+			ScriptLanguage::CodeActionOperation del_action;
+			ScriptLanguage::TextEdit del_op;
+			del_op.start_line = p_variable->start_line;
+			del_op.start_column = 1;
+			del_op.end_line = p_variable->start_line + 1;
+			del_op.end_column = 1;
+			del_op.new_text = "";
+			del_action.description = "Remove variable declaration";
```
Variable declarations can be across multiple lines, but this only takes the first line.
Eg.
```gdscript
	var unused2 = [
		5
	]
```

This probably applies to some other remove actions too.

Meorge replied:
Thanks for catching this! I've updated the code to use the full range of the node, so that it deletes the whole declaration for variables, class variables, and constants.

kitbdev commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -2266,7 +2466,39 @@ void GDScriptAnalyzer::resolve_variable(GDScriptParser::VariableNode *p_variable
+			del_op.end_line = p_variable->start_line + 1;
+			del_op.end_column = 1;
+			del_op.new_text = "";
+			del_action.description = "Remove variable declaration";
```
Unused variables can be assigned and still be unused, so any of those references won't be removed when the declaration is removed. It causes them to be an undeclared identifier error.

```gdscript
var unused = 1
unused = 2
```

Not sure if this can be fixed to delete all assignments too. Especially because it could be multi-document.

Meorge replied:
That's a good point that hadn't occurred to me before. I'm feeling inclined to just remove this code action from the PR. Even if there was a way to fix it to delete all assignments, I could see the possibility of it causing cascading errors if the variable is, for example, assigned the result of a function with side effects:
```gdscript

var health = 5

func take_damage():
  health -= 1
  return health

func _on_hit():
  var unused = 1
  unused = take_damage()
  print("new health is %s" % health)
```
This is a bit of a silly quick example, but hopefully the issue is still visible: If the code action removes `unused`'s assignment line, then the behavior of the overall program will be changed.

If you think it'd be best to just remove those code actions for now, I can get that done. The class variables and constants may also need it removed, as they can be assigned the results of functions with side effects too.

kitbdev replied:
The action does only say it will remove the *declaration*, so I think it is probably fine to keep it. It is still useful in common scenarios. 
And if the declaration has side effects, it is up to the user to decide to remove it or take another action for the warning.

Meorge replied:
Sounds good, I will keep it in as-is then 😄

kitbdev reviewed (approved):
Tested in ScriptEditor and external editor over LSP.
Editor code looks good.

Meorge commented:
Thank you very much for the review and approval! 😄

Repiteo commented:
Thanks!

Ivorforce commented:
This is missing a GDScript team review and was accidentally merged (see https://github.com/godotengine/godot/pull/121337).
@Meorge could you set up a new PR? Sorry for the hassle.

Meorge commented:
No worries, I will get that submitted in just a few minutes!
